pkg/storage/parameterset.go

```diff
@@ -84,21 +85,63 @@ func (s ParameterSet) DefaultDocumentFilter() map[string]interface{} {
 	return map[string]interface{}{"namespace": s.Namespace, "name": s.Name}
 }
 
-func (s *ParameterSet) Validate() error {
+func (s *ParameterSet) Validate(ctx context.Context, strategy schema.CheckStrategy) error {
+	ctx, span := tracing.StartSpan(ctx,
+		attribute.String("parameterSet", s.String()),
+		attribute.String("schemaVersion", string(s.SchemaVersion)),
+		attribute.String("defaultSchemaVersion", string(DefaultParameterSetSchemaVersion)))
+	defer span.EndSpan()
+
+	// Before we can validate, get our resource in a consistent state
+	// 1. Check if we know what to do with this version of the resource
+	if warnOnly, err := schema.ValidateSchemaVersion(strategy, SupportedParameterSetSchemaVersions, string(s.SchemaVersion), DefaultParameterSetSemverSchemaVersion); err != nil {
+		if warnOnly {
+			span.Warn(err.Error())
+		} else {
+			return span.Error(err)
+		}
+	}
+
+	// 2. Check if they passed in the right resource type
+	if s.SchemaType != "" && !strings.EqualFold(s.SchemaType, SchemaTypeParameterSet) {
+		return span.Errorf("invalid schemaType %s, expected %s", s.SchemaType, SchemaTypeParameterSet)
+	}
+
+	// 3. Migrate their resource to the most recently supported version if necessary
+	if err := s.Migrate(ctx); err != nil {
+		return err
+	}
+
+	// Default the schemaType before importing into the database if it's not set already
+	// SchemaType isn't really used by our code, it's a type hint for editors, but this will ensure we are consistent in our persisted documents
 	if s.SchemaType == "" {
-		// Default the schema type before importing into the database if it's not set already
-		// SchemaType isn't really used by our code, it's a type hint for editors, but this will ensure we are consistent in our persisted documents
 		s.SchemaType = SchemaTypeParameterSet
-	} else if !strings.EqualFold(s.SchemaType, SchemaTypeParameterSet) {
-		return fmt.Errorf("invalid schemaType %s, expected %s", s.SchemaType, SchemaTypeParameterSet)
 	}
 
-	if ParameterSetSchemaVersion != s.SchemaVersion {
-		if s.SchemaVersion == "" {
-			s.SchemaVersion = "(none)"
+	// OK! Now we can do resource specific validations
+	return nil
+}
+
+// Migrate checks if the resources is in an older schema version, and if so applies successive migrations until it is up-to-date.
+func (s *ParameterSet) Migrate(ctx context.Context) error {
+	//lint:ignore SA4006 ignore unused ctx for now
+	ctx, span := tracing.StartSpan(ctx)
+	defer span.EndSpan()
+
+	for s.SchemaVersion != DefaultParameterSetSchemaVersion {
+		switch s.SchemaVersion {
+		// In 1.0.2 we added SchemaType and any resources with it unset should have the default value applied
+		case "1.0.1":
+			span.Info("Migrating parameter set from 1.0.1 to 1.0.2")
+			// There isn't an explicit migration for this, because schemaType is funny
```

Comment:
I'm curious why you moved the code for setting schemaType if it's empty out of this `Migrate` function

Reply:
Originally you were concerned about inconsistently setting the schemaType, and having a hard time understanding when it's getting set. When I moved setting the schemaType only when migrating, I realized that only creates more opportunity for inconsistency.

The migration is only triggered if their schemaVersion is on an old version, which leaves open paths for people to still end up with the schemaType unset.

For example:

1. I have a resource with "schemaVersion: 1.0.1" and before applying it, see that I'm on an old schemaVersion, so I update the schemaVersion to 1.0.2.
2. I apply the resource, and since the version is "up-to-date", the schemaType is left empty.

or

1. I have a resource on 1.0.1 without the schemaType set.
2. I apply the resource and it's auto migrated to 1.0.2 and the type is set.
3. I keep using the original resource file, but perhaps realize that the version is out of date, and update it to 1.0.2 but otherwise keep the same file.
4. I apply it again, and now the schemaType in the database is unset.

or

1. I have a new resource that I made by hand. I set the schemaVersion to 1.0.2 but forget to set the schemaType
2. I apply it and the schemaType is empty in the database (because no migration was applied).

If the concern that drove us to set the value in Validate() was consistency, only defaulting in the migration doesn't solve the consistency problem. I originally wasn't concerned about if schemaType is set in the db, which is why I left it off but was willing to add the defaulting to address the consistency concern. So we should either ensure porter sets it consistently or just not bother and go back to the original because this was a lot of changes to remain inconsistent. 😁

Reply:
Thank you for the clarification!